src/createApp.ts

```diff
@@ -402,35 +453,51 @@ const createApp = compose({
 
 			return promise;
 		});
+		const instanceHandle = (<App> this)._instanceRegistry.addAction(action, id);
 
 		return {
 			destroy() {
 				this.destroy = noop;
+				instanceHandle.destroy();
 				registryHandle.destroy();
 			}
 		};
 	},
 
 	registerActionFactory(id: Identifier, factory: ActionFactory): Handle {
+		let destroyed = false;
+		let instanceHandle: Handle;
 		let registryHandle = actions.get(this).register(id, () => {
 			const promise = Promise.resolve().then(() => {
 				// Always call the factory in a future turn. This harmonizes behavior regardless of whether the
 				// factory is registered through this method or loaded from a definition.
 				return factory(this._registry);
+			}).then((action) => {
+				if (!destroyed) {
+					instanceHandle = (<App> this)._instanceRegistry.addAction(action, id);
```

Comment:
I would prefer an `const app: App = this;` at the top of the scope.  This will make it easier to refactor when `this` typing is allowed in TS2.  Also would eliminate the extra casts.